src/deploy/functions/release.ts

```diff
@@ -2,477 +2,120 @@
-  // If not using function filters, then `deleteReleaseNames` should be equivalent to existingNames so that intersection is a noop
-  const deleteReleaseNames = functionFilterGroups.length ? releaseNames : existingNames;
-  helper.logFilters(existingNames, releaseNames, functionFilterGroups);
-
-  const defaultEnvVariables = {
-    FIREBASE_CONFIG: JSON.stringify(context.firebaseConfig),
+  const fullDeployment = createDeploymentPlan(
+    payload.functions.byRegion,
+    context.existingFunctions,
+    context.filters
+  );
+  const cloudFunctionsQueue = new Queue<() => Promise<CloudFunctionTrigger | void>, void>({});
+  const schedulerQueue = new Queue<() => Promise<any>, void>({});
+  const regionPromises = [];
+
+  const taskParams: tasks.TaskParams = {
+    projectId,
+    sourceUrl,
+    runtime: context.runtimeChoice,
+    timer,
+    errorHandler,
   };
 
-  // Create functions
-  _.chain(uploadedNames)
-    .difference(existingNames)
-    .intersection(releaseNames)
-    .forEach((name) => {
-      const functionInfo = _.find(functionsInfo, { name: name })!;
-      const functionTrigger = helper.getFunctionTrigger(functionInfo);
-      const functionName = helper.getFunctionName(name);
-      const region = helper.getRegion(name);
-      utils.logBullet(
-        clc.bold.cyan("functions: ") +
-          "creating " +
-          getHumanFriendlyRuntimeName(runtime) +
-          " function " +
-          clc.bold(helper.getFunctionLabel(name)) +
-          "..."
-      );
-      logger.debug("Trigger is: ", JSON.stringify(functionTrigger));
-      const eventType = functionTrigger.eventTrigger
-        ? functionTrigger.eventTrigger.eventType
-        : "https";
-      startTimer(name, "create");
-      const retryFunction = async () => {
-        const createRes = await cloudfunctions.createFunction({
-          projectId: projectId,
-          region: region,
-          eventType: eventType,
-          functionName: functionName,
-          entryPoint: functionInfo.entryPoint,
-          trigger: functionTrigger,
-          labels: _.assign({}, deploymentTool.labels(), functionInfo.labels),
-          sourceUploadUrl: sourceUrl,
-          runtime: runtime,
-          availableMemoryMb: functionInfo.availableMemoryMb,
-          timeout: functionInfo.timeout,
-          maxInstances: functionInfo.maxInstances,
-          environmentVariables: defaultEnvVariables,
-          vpcConnector: functionInfo.vpcConnector,
-          vpcConnectorEgressSettings: functionInfo.vpcConnectorEgressSettings,
-          serviceAccountEmail: functionInfo.serviceAccountEmail,
+  const shouldDeleteFunctions = await promptForFunctionDeletion(
+    fullDeployment.functionsToDelete,
+    options.force,
+    options.nonInteractive
+  );
+  if (shouldDeleteFunctions) {
+    for (const fnName of fullDeployment.functionsToDelete) {
+      cloudFunctionsQueue
+        .run(tasks.deleteFunctionTask(taskParams, fnName))
+        .then(() => {
```

Comment:
Figured out a better way to do this error handling - another commit coming to refactor this slightly.